src/client/renderingengine.cpp

```diff
@@ -683,12 +696,42 @@ float RenderingEngine::getDisplayDensity()
 	return cached_display_density;
 }
 
-#else  // XORG_USED
+#elif defined(_WIN32)
+
+
+static float calcDisplayDensity(irr::video::IVideoDriver *driver)
+{
+	HWND hWnd;
+	if (getWindowHandle(driver, hWnd)) {
+		HDC hdc = GetDC(hWnd);
+		float dpi = GetDeviceCaps(hdc, LOGPIXELSX);
+		ReleaseDC(hWnd, hdc);
+		return dpi / 96.0f;
+	}
+
+	/* return manually specified dpi */
+	return g_settings->getFloat("screen_dpi") / 96.0f;
+}
+
+float RenderingEngine::getDisplayDensity()
```

Comment:
i don't think we should cache this, what happen when you change minetest window to another screen ?

Reply:
or we may have a function to detect the screen is not the old one and retrigger the calculation

Reply:
Runtime DPI changes require a [Window 10-only API](https://docs.microsoft.com/de-de/windows/win32/api/windef/ne-windef-dpi_awareness) and probably also needs to be implemented in Irrlicht. (and our GUI code might not be laid out for it)

Other platforms do not support that yet either.

Reply:
i don't know for GL rendering but on my work laptop i have 4k DPI and external screens and gnome apps are managing it properly. If you don't think it's possible, just ignore my comment

Reply:
It's definitely possible, but it should be implemented some other time in a different PR.

Reply:
okay fine to me